packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2025, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React, {useState, useEffect, useRef} from 'react'
+import {useBlockNavigation, useRoutes} from '@salesforce/pwa-kit-react-sdk/ssr/universal/hooks'
+import {useUrlMapping} from '@salesforce/commerce-sdk-react'
+import {useLocation, Redirect} from 'react-router-dom'
+import {useApplicationExtensionsStore} from '@salesforce/pwa-kit-extension-sdk/react'
+import {useExtensionConfig} from '../hooks/use-extension-config'
+type SeoHOCProps = React.ComponentPropsWithoutRef<any>
+
+interface UrlMappingResponse {
+    resourceType?: string
+    resourceId?: string
+    destinationUrl?: string
+}
+
+const getComponent = (
+    routes: Array<{component: React.ComponentType<any> | undefined; path: string}>,
+    resourceTypeToComponentMap: {[key: string]: string},
+    resourceType: string
+) => {
+    const ComponentClass = routes.find((_route) =>
+        _route.component?.displayName?.includes(resourceTypeToComponentMap[resourceType])
+    )?.component
+    return ComponentClass
+}
+
+const seoHOC = <P extends object>(WrappedComponent: React.ComponentType<P>) => {
+    const SeoHOC: React.FC<P> = (props: SeoHOCProps) => {
+        const location = useLocation()
+        const {routes, setRoutes} = useRoutes()
+        const {resourceTypeToComponentMap} = useExtensionConfig()
+        const [urlSegment, setUrlSegment] = useState(location.pathname)
+        const {setIsNavigationBlocked, siteLocale} = useApplicationExtensionsStore((state) => {
+            return state.state['@salesforce/extension-commerce-bm-seo']
+        })
+
+        const resolveRef = useRef<(result?: object) => void>()
+
+        const {refetch} = useUrlMapping(
+            {
+                parameters: {
+                    urlSegment: urlSegment,
+                    locale: siteLocale
+                }
+            },
+            {
+                enabled: false
+            }
+        )
+
+        useEffect(() => {
+            const fetchData = async () => {
+                if (urlSegment) {
+                    const result = await refetch()
+                    if (resolveRef.current) {
+                        if (result.status === 'error') {
+                            resolveRef.current(undefined)
+                        } else {
+                            if (result.data?.destinationUrl) {
+                                resolveRef.current(result.data)
+                            } else {
+                                resolveRef.current(undefined)
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
To be more readable, it'd be great if we flatten the nested conditions. Something like this:

```suggestion
            const fetchData = async () => {
                if (!urlSegment || !resolveRef.current) {
                    return
                }
                const result = await refetch()

                if (result.status === 'error') {
                    resolveRef.current(undefined)
                    return
                }

                const hasDestinationUrl = result.data?.destinationUrl
                resolveRef.current(hasDestinationUrl ? result.data : undefined)
            }
```